tools/src/testcert-gen/main.go

```diff
@@ -0,0 +1,302 @@
+		if err != nil {
+			return fmt.Errorf("read CA key: %w", err)
+		}
+		caKeyPEM, _ := pem.Decode(caKeyPEMBytes)
+		if caKeyPEM == nil {
+			return fmt.Errorf("decode CA key")
+		}
+		_caKey, err := x509.ParsePKCS8PrivateKey(caKeyPEM.Bytes)
+		if err != nil {
+			return fmt.Errorf("parse CA key: %w", err)
+		}
+		var ok bool
+		if caKey, ok = _caKey.(*rsa.PrivateKey); !ok {
+			return fmt.Errorf("CA key is not an RSA key: %w", err)
+		}
+	}
+
+	// Normally, you'd just use `crypto/rand.Reader`, but since we do this in `make generate`,
+	// we want it to be determistic (I named this `testcert-gen`, not `cert-gen`, because you
+	// should absoulutely never ever use this to generate a real cert.)
+	randReader := NewRand(name)
+
+	privKey, privKeyBytes, err := genKey(PrivArgs{
+		Rand: randReader,
+	})
+	if err != nil {
+		return err
+	}
+
+	pubBytes, err := genCert(CertArgs{
+		CACert: caCert,
+		CAKey:  caKey,
+
+		Rand: randReader,
+
+		CertKey:  privKey,
+		IsCA:     args.IsCA,
+		IsServer: args.IsServer,
+		IsClient: args.IsClient,
+		Hosts:    args.Hosts,
+	})
+	if err != nil {
+		return err
+	}
+
+	if err := ioutil.WriteFile(args.OutCert, pubBytes, 0666); err != nil {
+		return fmt.Errorf("writing cert to %q: %w", args.OutCert, err)
+	}
+	dlog.Printf(ctx, "%q: wrote cert to %q\n", name, args.OutCert)
+
+	if err := ioutil.WriteFile(args.OutKey, privKeyBytes, 0666); err != nil {
+		return fmt.Errorf("writing key to %q: %w", args.OutKey, err)
+	}
+	dlog.Printf(ctx, "%q: wrote key to %q\n", name, args.OutKey)
+
+	return nil
+}
+
+type PrivArgs struct {
+	Rand io.Reader
+}
+
+func genKey(args PrivArgs) (*rsa.PrivateKey, []byte, error) {
+	key, err := rsa.GenerateKey(args.Rand, 2048)
+	if err != nil {
+		return nil, nil, fmt.Errorf("generate key-pair: %w", err)
+	}
+
+	derBytes, err := x509.MarshalPKCS8PrivateKey(key)
+	if err != nil {
+		return nil, nil, fmt.Errorf("marshal private key: %w", err)
+	}
+
+	pemBytes := pem.EncodeToMemory(&pem.Block{
+		Type:  "PRIVATE KEY",
+		Bytes: derBytes,
+	})
+
+	return key, pemBytes, nil
+}
+
+type CertArgs struct {
+	CACert *x509.Certificate
+	CAKey  *rsa.PrivateKey
+
+	Rand io.Reader
+
+	CertKey  *rsa.PrivateKey
+	IsCA     bool
+	IsServer bool
+	IsClient bool
+	Hosts    []string
+}
+
+func genCert(args CertArgs) ([]byte, error) {
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, err := crypto_rand.Int(args.Rand, serialNumberLimit)
+	if err != nil {
+		return nil, fmt.Errorf("generate serial number: %w", err)
+	}
+
+	template := x509.Certificate{
+		SerialNumber: serialNumber,
+		Subject: pkix.Name{
+			Country:            []string{"US"},
+			Province:           []string{"MA"},
+			Locality:           []string{"Boston"},
+			Organization:       []string{"Ambassador Labs"},
+			OrganizationalUnit: []string{"Engineering"},
+
+			CommonName: args.Hosts[0],
+		},
+		NotBefore: time.Date(2021, 11, 10, 13, 12, 0, 0, time.UTC),
+		NotAfter:  time.Date(2099, 11, 10, 13, 12, 0, 0, time.UTC),
```

Comment:
This will give MacOS browsers fits. I'm pretty sure we don't care – is there ever a reason we'd want to hit a server using these certs from a browser?

Reply:
We don't care. I'll add a comment saying that.